x/ibc-rate-limit/contracts/rate-limiter/src/contract.rs

```diff
@@ -0,0 +1,520 @@
+) -> Result<Response, ContractError> {
+    // Only the IBCMODULE can execute transfers
+    // TODO: Should we move this to a helper method?
+    let ibc_module = IBCMODULE.load(deps.storage)?;
+    if sender != ibc_module {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let channels = CHANNEL_FLOWS.may_load(deps.storage, &channel_id)?;
+
+    let configured = match channels {
+        None => false,
+        Some(ref x) if x.is_empty() => false,
+        _ => true,
+    };
+
+    if !configured {
+        // No Quota configured for the current channel. Allowing all messages.
+        // TODO: Should there be an attribute for it being allowed vs denied?
+        return Ok(Response::new()
+            .add_attribute("method", "try_transfer")
+            .add_attribute("channel_id", channel_id)
+            .add_attribute("quota", "none"));
+    }
+
+    let mut channels = channels.unwrap();
+
+    let results: Result<Vec<ChannelFlowResponse>, _> = channels
+        .iter_mut()
+        .map(|channel| {
+            // TODO: Should we move this to more methods on ChannelFlow?
+            // e.g. new pseudocode
+            // channel.updateIfExpired(now)
+            // channel.trackSend(&direction, funds)
+            // channel.CheckRateLimit(direction.clone())?;
+            // (Or at least update for time + rename for track send. the last one is a bit of a code style nit)
+            let max = channel.quota.capacity_at(&channel_value, &direction);
+            if channel.flow.is_expired(now) {
+                channel.flow.expire(now, channel.quota.duration)
+            }
+            channel.flow.add_flow(direction.clone(), funds);
+
+            let balance = channel.flow.balance();
+            if balance > max {
+                return Err(ContractError::RateLimitExceded {
+                    channel: channel_id.to_string(),
+                    reset: channel.flow.period_end,
+                });
+            };
+            Ok(ChannelFlowResponse {
+                // TODO: nit, can we derive channel.Clone()?
+                channel_flow: ChannelFlow {
+                    quota: channel.quota.clone(),
+                    flow: channel.flow,
+                },
+                used: balance,
+                max,
+            })
+        })
+        .collect();
+    let results = results?;
+
+    CHANNEL_FLOWS.save(
+        deps.storage,
+        &channel_id,
+        &results.iter().map(|r| r.channel_flow.clone()).collect(),
+    )?;
+
+    let response = Response::new()
+        .add_attribute("method", "try_transfer")
+        .add_attribute("channel_id", channel_id);
+
+    // Adding the attributes from each quota to the response
+    // Code style Q: Should we move attribute setting to a function on response?
+    // Rust question: How does this work with one response being an error, I'm not sure how the flow works here
+    results.iter().fold(Ok(response), |acc, result| {
+        Ok(acc?
```

Comment:
@ValarDragon for the rust question: the `?` here on the accumulator catches and returns any potential error

Reply:
Did some reorg here. We have a function to set the attributes, and I'm considering making those attributes only available in testing. They are, after all, only seen by the state machine, not the end-user (and the response is being discarded).